src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs

```diff
@@ -266,7 +267,19 @@ protected override void OnScriptFileChanged(object sender, FileSystemEventArgs e
                     {
                         TraceLevel level = (TraceLevel)logData["lvl"];
                         var evt = new TraceEvent(level, message);
-                        traceWriter.Trace(evt);
+
+                        // Node captures the AsyncLocal value of the first invocation, which means that logs
+                        // are correlated incorrectly. Here we'll overwrite that value with the correct value
+                        // immediately before logging.
+                        using (logger.BeginScope(
+                            new Dictionary<string, object>
+                            {
+                                ["MS_FunctionInvocationId"] = invocationContext.ExecutionContext.InvocationId
+                            }))
+                        {
+                            // TraceWriter already logs to ILogger
+                            traceWriter.Trace(evt);
```

Comment:
I mostly agree with you. We can revisit this -- at first I wanted to keep everything separate so it was clearer but I think both @paulbatum and @christopheranderson wanted users of TraceWriters to automatically get App Insights. And since this TraceWriter is coming from that same TraceWriterBinding, we'd need a way to "undo" that, which also gets messy.

I've got my fingers crossed that we can remove TraceWriter completely soon and then all of this confusion starts to go away -- but that's a separate discussion :-)